Support running jasmine multiple times

nicojs opened the pull request:
Add support for running jasmine multiple times.

```js
const Jasmine = require('jasmine');

async function main() {
  const jasmine = new Jasmine({ projectBaseDir: process.cwd() });
  let specId = 'spec0';
  jasmine.loadConfigFile('./spec/support/jasmine.json');
  jasmine.env.configure({
    specFilter(sp) {
      return sp.id === specId;
    },
    autoCleanClosures: false
  });
  jasmine.exit = () => {};
  await jasmine.execute();
  specId = 'spec2';
  await jasmine.execute();
}

main().catch((err) => {
  console.error(err);
  process.exitCode = 1;
});
```

With `jasmine.env.configure({ autoCleanClosures: false })` you disable Jasmine's feature to automatically clean closures (functions) during the test run. This is a requirement to be able to rerun.

When `execute` is called more than once, the `topSuite.reset` is called, which will reset the state for the next run as well as reset any child suites.

Add a function `exclude` to the `Suite` and `Spec` clases. This functions similar to `pend`, but will allow the "pending" state to persist over multiple runs. This is useful when `xit` is used.

<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

nicojs commented:
@sgravrock as discussed in #1925. I think I didn't break anything.

Q: Where do I update the docs?

sgravrock commented:
Thanks for the PR. I hope to have time to give it a deeper look in the next few days.

> Q: Where do I update the docs?

If it's just a matter of API reference, you can add jsdoc comments (looks like you've already done that) and I'll update the docs site after merging the PR. If you think something more extensive like a tutorial is needed, a PR against https://github.com/jasmine/jasmine.github.io/ would be welcome.

sgravrock commented:
The code looks reasonable based on a quick-ish skim, but I can't get it to work. I ran `npm run build`, applied the below patch to your branch, and ran `npm test`. I think should cause Jasmine's own test suite to run twice using the new feature. The first run passes but the second run produces a variable number of spec failures, many of them async timeouts.

```
diff --git a/Gruntfile.js b/Gruntfile.js
index 14077d78..13f18d65 100644
--- a/Gruntfile.js
+++ b/Gruntfile.js
@@ -40,8 +40,17 @@ module.exports = function(grunt) {
           jasmine = new Jasmine({jasmineCore: jasmineCore});
 
       jasmine.loadConfigFile('./spec/support/jasmine.json');
+      jasmine.env.configure({autoCleanClosures: false});
+      var firstRunPassed;
       jasmine.onComplete(function(passed) {
-        done(passed);
+        console.log(">>>>>> run finished:", passed);
+        if (firstRunPassed === undefined) {
+          firstRunPassed = passed;
+          console.log(">>>>>> starting 2nd run");
+          jasmine.execute();
+        } else {
+          done(passed && firstRunPassed);
+        }
       });
 
       jasmine.execute();

```

It also looks like the onComplete callback is called twice on the second run, for a total of three times.

On a more minor note, the comments you added to Spec#pend, Spec#exclude, Suite#pend, and Suite#exclude are helpful but they shouldn't use jsdoc syntax (`/**`) because those methods shouldn't be part of the public interface.

nicojs commented:
Great idea with regards to the patch. I was looking for easy ways of performing e2e tests but didn't think of this one 🤦‍♀️ I'll have a look!

I will also update the comments and fix the conflicts with the main branch 👍

nicojs commented:
@sgravrock I found the issue and fixed it. Interesting bug. Had to do with function scoped variables, which resulted in before/after hooks not being executed in the second run. I've fixed it and added an integration test.

Your grunt patch now works:

```
Started
..........................................................................................................................................................................................................................................................................*..................................................................................................................................................................................................................................................................................................................................................................................................................................................................*...............................................................................................................................................................................................................................................................................................................................................................................................................................*.......................................................................................................................................................................................................................................................................

Pending:

1) Env #execute returns undefined when promises are unavailable
  Environment supports promises


2) base helpers isError_ returns true for an Error that originated from another frame
  This test only runs in browsers.


3) AsyncExpectation #withContext prepends the context to a custom failure message from a function
  should actually work, but no custom matchers for async yet

1397 specs, 0 failures, 3 pending specs
Finished in 3.658 seconds
Randomized with seed 98449 (jasmine --random=true --seed=98449)
>>>>>> run finished: true
>>>>>> starting 2nd run
Randomized with seed 89143
Started
...............................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................*................................................................................................................................................................*...............................................................................................................................................................................................................................................................................................................................*....................................

```

I've also removed the JS doc (made it regular comments)

> It also looks like the onComplete callback is called twice on the second run, for a total of three times.

I've noticed that as well. Unfortunately, nothing I can do for that this PR. Needs to be fixed in [jasmine-npm](https://github.com/jasmine/jasmine-npm) after this PR is merged. The completion reporter shouldn't be added the second time around: https://github.com/jasmine/jasmine-npm/blob/18098bfc976e3f9272c94ef279c1466325c6520d/lib/jasmine.js#L562

I'd be happy to present the PR there after this PR is merged.

sgravrock commented:
Looks good. I think this is just about ready to merge. Could you put blank lines between the new specs in SuiteSpec.js and squash your commits? Or I can squash for you if you're not comfortable doing it, but in that case GitHub won't show the PR as merged.

nicojs commented:
👍 did both

nicojs commented:
Hmm the build is red.

Is there a way to view the circleci logs without logging in? I don't feel comfortable providing them with so much power
![image](https://user-images.githubusercontent.com/1828233/136182765-eadc7ee4-1bc0-4909-b9f3-a56eb1bbf61d.png)

sgravrock commented:
```
1397 specs, 0 failures, 3 pending specs
Finished in 3.938 seconds
Randomized with seed 69323 (jasmine --random=true --seed=69323)

Done.

> jasmine-core@3.9.0 posttest /home/circleci/workspace
> eslint "src/**/*.js" "spec/**/*.js" && prettier --check "src/**/*.js" "spec/**/*.js"

Checking formatting...
spec/core/SuiteSpec.js
Code style issues found in the above file(s). Forgot to run Prettier?
npm ERR! Test failed.  See above for more details.
```

If you run `npm run cleanup` and amend your commit with the changes, that should fix it.

nicojs commented:
Ah facepalm moment 🤦‍♀️. Fixed now.

nicojs commented:
Thanks! Will take care of it tomorrow
